src/parser/parser_utils.rs

```diff
@@ -0,0 +1,278 @@
+/// amount parameter specifies number of nodes/tokens to kill
+pub fn kill_node_attribute(node: &NixNode, amount: usize) -> anyhow::Result<NixNode> {
+    let parent = node.parent().unwrap();
+    match parent.kind() {
+        NODE_ATTR_SET | NODE_PATTERN => {
+            let mut child_node_idxs =
+                parent
+                    .green()
+                    .children()
+                    .enumerate()
+                    .filter_map(|(idx, val)| {
+                        // the '.to_owned()' is required to turn GreenNodeData into GreenNode
+                        // because GreenNodeData doesn't implement PartialEq
+                        val.into_node().and_then(|inner_node| {
+                            (*inner_node == node.green().to_owned()).then(|| idx)
+                        })
+                    });
+            // only one child node should match
+            let idx = child_node_idxs.next().expect("Child not in parent tree");
+            assert!(
+                child_node_idxs.next().is_none(),
+                "AST in inconsistent state. Child found multiple times in parent tree."
+            );
+            let result = node
+                .parent()
+                .unwrap()
+                .green()
+                .splice_children(idx..idx + amount, std::iter::empty());
+            let mut new_root = NixNode::new_root(parent.replace_with(result));
+            while let Some(parent) = new_root.parent() {
+                new_root = parent;
+            }
+            let tmp = Root::cast(new_root).unwrap();
+            Ok(tmp.inner().unwrap())
+        }
+        _ => Err(anyhow!(
+            "Precondition violated: parent was not attribute set.".to_string()
+        )),
+    }
+}
+
+/// converts a AST node to a string.
+pub fn node_to_string(node: &NixNode) -> String {
+    Str::cast(node.clone())
+        .unwrap()
+        .parts()
+        .iter()
+        .fold(String::new(), |mut acc, ele| -> String {
+            match ele {
+                StrPart::Literal(s) => {
+                    acc.push_str(s);
+                    acc
+                }
+                StrPart::Ast(_) => unimplemented!(),
+            }
+        })
+}
+
+pub fn string_to_node(content: String) -> anyhow::Result<NixNode> {
+    let ast = match rnix::parse(&content).as_result() {
+        Ok(parsed) => parsed,
+        Err(err) => {
+            bail!(format!("could not parse as a nix file: {}", err));
+        }
+    };
+    Ok(ast.root().inner().unwrap())
+}
+
+/// given an attribute name, searches to max_depth
+/// for the given attribute name
+/// it is assumed that the node's max depth >= exact_depth
+/// and root_node is of type attrset
+/// returns a vector of tuples that match
+/// (matching_node, path, depth)
+/// Example: searching for `foo`
+/// ```
+/// "{\"foo\": \"bar\"}"
+/// ```
+/// Will return [bar_node, "foo", 1]
+fn search_for_attr(
+    attr: String,
```

Comment:
consider change this attribute to `&str`. In most call-sites, a basically useless `"...".to_string()` is used to turn a `&'static str` into an owned `String`, which is unnecessary, slow, and clutters the code.
```suggestion
    attr: &str,
```
when applying, the `search_for_attr("$whatever".to_string(),` stuff should be changed to `search_for_attr("$whatever",`... maybe, at some call-sites, it may be necessary to insert a `&` where previously an owned `String` was passed.